### ISSUE TITLE
ERROR - RequirementParseError: Invalid requirement, parse error at "’'"

### ISSUE THREAD

**Issue by dsethi** [paramiko/paramiko#931]
I am getting the following issue, with paramiko 2.0.2 and multi-threading from Pool in multiprocessor.dummy: 
ERROR - Unknown exception: Invalid requirement, parse error at "''"
ERROR - Traceback (most recent call last):
ERROR -   File "/usr/lib/python2.7/site-packages/paramiko/transport.py", line 1774, in run
ERROR -     self.kex_engine.parse_next(ptype, m)
ERROR -     return self._parse_kexdh_reply(m)
 ERROR -   File "/usr/lib/python2.7/site-packages/paramiko/kex_group1.py", line 111, in _parse_kexdh_reply
ERROR -     self.transport._verify_key(host_key, sig)
ERROR -   File "/usr/lib/python2.7/site-packages/paramiko/transport.py", line 1619, in _verify_key
ERROR -     key = self._key_info[self.host_key_type](Message(host_key))
ERROR -   File "/usr/lib/python2.7/site-packages/paramiko/rsakey.py", line 58, in __init__
ERROR -     ).public_key(default_backend())
ERROR -   File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/backends/__init__.py", line 35, in default_backend
ERROR -     _default_backend = MultiBackend(_available_backends())
ERROR -   File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/backends/__init__.py", line 22, in _available_backends
ERROR -     “cryptography.backends"
 ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 706, in iter_entry_points
ERROR -     entries = dist.get_entry_map(group)
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2578, in get_entry_map
ERROR -     self._get_metadata('entry_points.txt'), self
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2316, in parse_map
ERROR -     maps[group] = cls.parse_group(group, lines, dist)
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2294, in parse_group
ERROR -     ep = cls.parse(line, dist)
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2276, in parse
ERROR -     return cls(res['name'], res['module'], attrs, extras, dist)
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2203, in __init__
ERROR -     self.extras = Requirement.parse(("x[%s]" % ','.join(extras))).extras
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2858, in parse
ERROR -     req, = parse_requirements(s)
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2806, in parse_requirements
ERROR -     yield Requirement(line)
ERROR -   File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2815, in __init__
ERROR -     raise RequirementParseError(str(e))
ERROR - RequirementParseError: Invalid requirement, parse error at "’'"

**Comment by bitprophet** [comment]
Haven't seen that on my end but a quick Google implies this may be an issue with setuptools (unfortunately, despite all the awesome work they are doing on that lately, it seems a little unstable right now). I would try downgrading your setuptools and/or pip and see if that helps.

For example, `pip install "setuptools<34" will get the previous (bit more stable) release line. Good luck - can't give many specifics unfortunately.

**Comment by flowbish** [comment]
I was running into this exact same issue, but changing the version of setuptools I was using didn't seem to help at all. I ended up resorting to globally locking around client `connect()` calls which is super annoying.

So, I did a bit of digging and it seems like this is caused by the behavior of cryptography to support multiple backends using pkg_resources's entry_points API. Cryptography 1.9, released two days ago, completely removes this multiple-backend behavior and with that this issue goes away. I'm not quite sure why using entry_points from multiple threads was causing issues -- admittedly I couldn't repro this myself with pkg_resources alone, but using paramiko I could consistently get the same backtrace.

I just wanted to leave this comment here to document what I've found. Thanks!
